### DISCUSSION TITLE
Forcing antialiasing for native d3d9 game?

### DISCUSSION THREAD

**Discussion by TerrorizedGakon** [elishacloud/dxwrapper#133] [General]
Hi,

Can dxwrapper be used for this purpose?

I want to use MSAA in Gothic 3, but since I have AMD graphics card, it seems impossible to force it from the AMD panel.

I tried using dgvoodoo in the past, which worked, but caused game to sometimes crash on save loading.

**Comment by erwingame** [comment]
You can try it.

**Comment by elishacloud** [comment]
Yes, it should be possible to enable antialiasing in native DirectX 9 games.  However, it doesn't work for all games.  To do this you will need to enable the following settings in the ini file:

```ini
EnableD3d9Wrapper          = 0
AntiAliasing               = 0
```

**Comment by mirh** [comment]
> but since I have AMD graphics card, it seems impossible to force it from the AMD panel.

Uhm? Just enable it from the control panel and you are good (assuming the thing doesn't eventually glitch out)

**Comment by elishacloud** [comment]
Forcing antialiasing from the control panel does not work on all games.  For example it does nothing on Silent Hill 2.  I had to make a custom feature to allow GPU antialiasing to work at all, and even at that it only works on some GPUs, not all.  Furthermore, forcing GPU antialiasing does not work on all video cards.  Some of the older or some integrated cards don't have an option for antialiasing.

**Reply by mirh** [reply]
Mhh.. I [see](https://forums.guru3d.com/threads/nvidia-anti-aliasing-guide-updated.357956/page-170) even [back](https://www.hwupgrade.it/forum/showthread.php?t=450369) in the days it just used to work with nvidia cards.

**Reply by elishacloud** [reply]
Yes, I forgot GPU-forced antialiasing works for the most part in Silent Hill 2.  However, I believe there are some places in Silent Hill 2, like room 213 (I think it is) that does not work without the [fix I added](https://github.com/elishacloud/Silent-Hill-2-Enhancements/blob/d4f76a42e739d8efe7a43882c15e5e46b9007938/Common/Settings.ini#L318-L319).

Also, there are games that don't work at all with GPU forced antialiasing.  [Serious Sam](https://github.com/crosire/d3d8to9/issues/63) is another example game.  I had to [update](https://github.com/crosire/d3d8to9/issues/63#issuecomment-581078069) the d3d8to9 code to get antialiasing to work.  If games incorrectly call CreateDevice(), they may not be able to even support GPU forced antialiasing.

**Reply by mirh** [reply]
I thought it was only dx10+ games (forward rendering, amirite?) that were a total PITA to hack in. 
..
And I thought that SSAA still worked wherever?

**Reply by elishacloud** [reply]
I am not sure about dx10+ games.  I have not done any work on those versions of DirectX.

For SSAA, it doesn't work on games with lockable backbuffers or with games that use certain SwapEffects.  For all the requirements for SSAA see [here](https://docs.microsoft.com/en-us/windows/win32/direct3d9/full-scene-antialiasing).

As far as GPU forced antialiasing (using the control panel), I am not sure exactly what the requirements are.  But I know that some games won't work with it, I have seen it myself.

**Reply by mirh** [reply]
I guess mass effect having lights with a negative "showing" bias could also count to those problems.
